### ISSUE TITLE
Examining state of the current notebook

### ISSUE THREAD

**Issue by pitosalas** [jupyterlab/jupyterlab#4395]
The one thing I constantly need when I use Jupyter notebook is a quick way to remember the column names of a pandas datatable, or a quick way to see the first few rows (.head()). Kind of an "inspector". I see such a command in Lab but I didn't see any documentation for it and for me it doesn't do anything that I can tell.

Is there a feature like what I describe in Lab or is it planned?

Thanks,

**Comment by ian-r-rose** [comment]
@pitosalas You can open a console that is attached to the same kernel as an active notebook, which can then be used as a kind of inspector/scratch space for that notebook. You can do this using File --> "New Console For Notebook".

Does that sound like it would solve the issue?

**Comment by jasongrout** [comment]
Note also that there also is an open issue for implementing a variable inspector: https://github.com/jupyterlab/jupyterlab/issues/443

**Comment by pitosalas** [comment]
Yes @ian-r-rose that would help. By the way, is Lab meant to be the future replacement for Notebook? They seem very similar?

**Comment by ian-r-rose** [comment]
Yes, Lab is intended to be the replacement for the Notebook, designed around being more flexible and extensible.

**Comment by BoPeng** [comment]
If you are still using Jupyter, [SoS Notebook](https://vatlab.github.io/sos-docs/) (a jupyter kernel for sos) provides a side panel that you can used to evaluate scripts outside of the main notebook and keep the information there.

**Comment by pitosalas** [comment]
Thanks everyone! While I wait for that feature to appear in Lab I decided to use Spyder.

Pito Salas
Brandeis Computer Science
Volen 134

**Comment by pitosalas** [comment]
> 
> 	1) Currently you can open a console connected to the same kernel as a notebook (File -> New Console for Notebook). You can arrange the console side-by-side with your notebook in order to inspect things in the kernel outside of the notebook. A problem (for your usecase) is that every code execution in the notebook is also sent to this console, so if you do a .head() and then execute some notebook cells, pretty soon the .head() has scrolled up out of view.

I hadn’t realized that things would scroll of. That would make this idea not so suitable.


> 	2) Another possibility is just creating a new notebook hooked to the same kernel. Then you have two notebooks hooked to the same kernel - in one you can do your normal work, in the other you can evaluate and keep side-by-side the helpful information you want. Disadvantage is now you have another file on your filesystem.

I don’t mind the file, but having to constantly type control-enter to re-evaluate the cell would not be ideal.

> 	3) @BoPeng's idea of a side panel is interesting - essentially a single notebook cell off to the side hooked to the same kernel as the running notebook. You can evaluate some code there once and the output stays there. This should be fairly straightforward to do (@BoPeng already has started a PR, IIRC).

That would be nice especially if you could somehow have it automatically trigger a re-eval. The benefit here is that it can be any valid statement not just the graphical display of a particular variable value

> 	4) Or are you really just wanting a variable inspector that updates after every computation, like Spyder has?

That would be nice, but it’s less powerful than (3)

> Or perhaps more than a variable inspector, a 'watch' pane that evaluates an expression after every execution (so you can have a dynamically updated .head()). That will be harder to develop, but I do see us having it sometime in the future, perhaps as a contribution from the community.

That is nice too, but again analogous to 3). Remember 3) could be generating a graph or doing anything. That sounds pretty nice.

> I see you teach at Brandeis. Do you know any students or others interested in a project to develop something like a variable inspector

I can’t really say. By the way, before coming here I was pursuing R and RStudio, and I also use Eclipse, so there are best practice examples that we can emulate.

**Comment by jasongrout** [comment]
> so there are best practice examples that we can emulate.

Are there any other examples you can elaborate on here from your experience?

**Comment by BoPeng** [comment]
Just to expand on 3),

1. The off-notebook in the side panel cell can accept direct input so you can evaluate anything there. This feature is similar to the console window of JLab but as @jasongrout pointed out, output of normal cells will be sent to the console as well so you cannot really keep the output you want in the console window. 
2. You can send selected code or current line to the side panel using shortcut `Ctrl-Shift-Enter`, essentially doing line-by-line execution. A PR is available for JupyterLab (https://github.com/jupyterlab/jupyterlab/pull/4330).
3. The cell is also used to display output of a [`preview` magic](https://vatlab.github.io/sos-docs/doc/documentation/Notebook_Interface.html#Preview-of-results--14) so you can put the magic in the main notebook to preview the result of cells in the side panel.
4. SoS also provides an auto-preview feature. That is to say, if you use Ctrl-Shift-Enter to step through a line with assignment (e.g. `a=1`), SoS will automatically preview `a`. This works even for non-Python cells (e.g. `a <- 1` in R).
5. Finally, although SoS is designed for multiple kernels in one notebook, you can set a global language and use it throughout the notebook, so you can use SoS even if you only use R.

I will create a separate ticket for point 1 because I think it is better not to send output of notebook to console window.

**Comment by pitosalas** [comment]
By the way, as you bring up SoS: I looked at the site and don’t really understand what it is. Is it just a better Jupiter notebook?

Pito Salas
Brandeis Computer Science
Volen 134

**Comment by BoPeng** [comment]
I really need to update the SoS website to better explain what SoS is. :smile:

SoS consists of SoS Polyglot Notebook and SoS workflow engine.
1. SoS Notebook is a Jupyter extension that allows the use of multiple jupyter kernels in one notebook, with data exchange among live kernels. Technically speaking, SoS notebook consists of a customized kernel that is a super kernel to other Jupyter kernels, with a number of magics to manage subkernels and facilitate data exchange between them. SoS Notebook also extends the Jupyter frontend with features like line-by-line execution.  [This blog post](https://vatlab.github.io/blog/post/sos-notebook/) explains features of SoS Notebook.
2. SoS workflow engine is a Python3 based workflow engine that tries to overcome the barrier of using workflows in daily computational research. [This blog post](https://vatlab.github.io/blog/post/sos-workflow-engine/) summarizes the design of SoS workflow engine.
3. Although SoS workflow engine can be used from command line, it is designed to be used in SoS Notebook so that users can analyze data or develop scripts interactively in SoS Notebook, and convert part or all steps to SoS workflows to analyze large amount of data. [Our third blog post](https://vatlab.github.io/blog/post/power-of-sos-plus-sos-notebook/) talks about the relationship between SoS Notebook and SoS workflow engine.

As far as this ticket is concerned, we are talking about the frontend extensions of SoS. If you use Jupyter, you can install the sos kernel, create notebooks with it, and use the frontend features without worrying out the workflow engine part. I am porting SoS Notebook to JupyterLab but instead of keeping these frontend extensions within an SoS extension, I am exploring the possibilities of making them general features of JupyterLab. Examples of these features include [line by line execution](https://github.com/jupyterlab/jupyterlab/pull/4330) and [the multi-purpose side panel](https://github.com/jupyterlab/jupyterlab/issues/4411).

**Comment by jasongrout** [comment]
Another approach: https://github.com/jupyter-widgets/jupyterlab-sidecar

Basically, it allows creating a sidepanel with arbitrary output, using the ipywidgets framework. For example:

```python
from sidecar import Sidecar
import pandas as pd
s = Sidecar(title='Debug')

```
```python
df = pd.DataFrame([{'a': 1, 'b': 2}, {'a': 5, 'b': 10, 'c': 20}])
with s:
    display(df)
```
```python
s.clear_output()
```
<img width="1106" alt="screen shot 2018-04-23 at 9 39 17 am" src="https://user-images.githubusercontent.com/192614/39140713-4e68b470-46da-11e8-8ebc-7c1396ea2533.png">

**Comment by jasongrout** [comment]
CC @SylvainCorlay

**Comment by BoPeng** [comment]
This sidecar is very similar to the side panel that I proposed in #4411, although some equivalence of [SoS' `%preview` magic](https://vatlab.github.io/sos-docs/doc/documentation/Notebook_Interface.html#Preview-of-results--14)  is needed to display such a table. I would argue that the sidepanel is more general and easier to use than sidecar though.

I have tried to implement #4411 over the weekend, hopefully I can produce some working prototype in the next few weeks.

**Comment by jasongrout** [comment]
To be clear, I didn't propose it as *the* approach to take, but as an approach that works today and can inform our discussion.

**Comment by BoPeng** [comment]
I understand. I originally thought that the new console window (compare to classic jupyter) would be a good destination for line-by-line execution (#4330) but the fact that console window also displays output of regular cells makes it less usable for this purpose (#4424). Given that executing some expressions, especially displaying tables for dataframes as shown in this thread can be a common use case for "inspection", I now tend to think it is better to send code from line-by-line execution to an inspection panel (#4411). It is a bit difficult to modify the inspector panel to accommodate this usage though.

**Comment by jasongrout** [comment]
An interesting composable approach for the ipython kernel would be to create an ipython magic to execute an arbitrary block of code before/after/around every execution (https://github.com/ipython/ipython/issues/11104). With a magic like that, you could easily execute a block of code that updates an output widget to display a variable after each execution, for example.

Downside is that it wouldn't be a language agnostic approach, which may be better done by having that update step initiated from the browser.

**Comment by jtkiley** [comment]
I would also find this really helpful, and here's an example of similar functionality in Stata. Note a few things:

1. Stata only has one dataset active at a time. For jupyterlab, you would probably want a way to specify which dataframe is being displayed. I don't know about others' workflows, but I tend to work on the same one for a bit before moving on (usually sequentially followed by a merge), so a dropdown wouldn't be an issue.
1. There's a search field for live filtering of names. That's really helpful, as I frequently have hundreds of columns.
1. Dataset metadata is at the bottom of that right side pane. I like that because I occasionally catch bugs (generally in merging logic) by noticing the row count change. It would be extra helpful with a pandas workflow, because I often push against RAM limits with pandas.
1. Currently, I tend to use Python/pandas/jlab for data prep and examination, then I move to Stata for analyses. That may change as statsmodels and linearmodels grow over time, and should I decide to rework the publication-ready table automation that I currently have in Stata.

<img width="1302" alt="stata_example" src="https://user-images.githubusercontent.com/3935135/40016736-e8e34440-577c-11e8-9759-86d6c4ea8adf.png">
